lib/src/util/condition_scope_visitor.dart

```diff
@@ -0,0 +1,259 @@
+          ? DartTypeUtilities
+              .getCanonicalElement(leftPart.propertyName.bestElement)
+          : null;
+}
+
+bool _isLoopStatement(AstNode node) => node is DoStatement ||
+      node is ForEachStatement ||
+      node is ForStatement ||
+      node is WhileStatement;
+
+_needsLocalEnvironment(AstNode node) =>
+    node is ClassDeclaration ||
+    node is CompilationUnit ||
+    node is ConstructorDeclaration ||
+    node is FunctionDeclaration ||
+    node is MethodDeclaration;
+
+abstract class ConditionScopeVisitor extends UnifyingAstVisitor {
+  final Queue<Queue<_ExpressionBox>> environments = new Queue();
+
+  Iterable<Expression> getFalseExpressions(Iterable<Element> elements) =>
+      _getExpressions(elements, value: false);
+
+  Iterable<Expression> getTrueExpressions(Iterable<Element> elements) =>
+      _getExpressions(elements);
+
+  @override
+  visitAssignmentExpression(AssignmentExpression node) {
+    _addElementToEnvironment(new _UndefinedExpression(_getLeftElement(node)));
+    node.visitChildren(this);
+  }
+
+  void visitCondition(Expression node);
+
+  @override
+  visitDoStatement(DoStatement node) {
+    visitCondition(node.condition);
+    bool hasBreak = DartTypeUtilities
+        .traverseNodesInDFS(node, excludeCriteria: (node) => _isLoopStatement(node))
+        .any((e) => e is BreakStatement);
+    if (!hasBreak) {
+      _addElementToEnvironment(
+          new _ConditionExpression(node.condition, value: false));
+    }
+  }
+
+  @override
+  visitForStatement(ForStatement node) {
+    _addLocalEnvironment();
+    _addElementToEnvironment(new _ConditionExpression(node.condition));
+    node.variables?.accept(this);
+    node.initialization?.accept(this);
+    visitCondition(node.condition);
+    node.condition?.accept(this);
+    _addElementToEnvironment(new _ConditionExpression(node.condition));
+    node.updaters.accept(this);
+    node.body?.accept(this);
+    _removeLocalEnvironment();
+    bool hasBreak = DartTypeUtilities
+        .traverseNodesInDFS(node, excludeCriteria: _isLoopStatement)
+        .any((e) => e is BreakStatement);
+    if (!hasBreak) {
+      _addElementToEnvironment(
+          new _ConditionExpression(node.condition, value: false));
+    }
+  }
+
+  @override
+  visitIfStatement(IfStatement node) {
+    visitCondition(node.condition);
+    final elseUndefinedExpressions =
+        _visitElseStatement(node.elseStatement, node.condition);
+    _visitIfStatement(node);
+    elseUndefinedExpressions
+        .where((e) => e is _UndefinedExpression)
+        .forEach(_addElementToEnvironment);
+    if (_isLastStatementAnExitStatement(node.thenStatement)) {
+      _addElementToEnvironment(
+          new _ConditionExpression(node.condition, value: false));
+    }
+    if (_isLastStatementAnExitStatement(node.elseStatement)) {
+      _addElementToEnvironment(new _ConditionExpression(node.condition));
+    }
+  }
+
+  @override
+  visitNode(AstNode node) {
```

Comment:
Does the overridden method do exactly what this one does? (I think it does, but it's hard for me to check right now.) If so, let's just delete this method.